File: 프로그래머스/js/Level 2/쿼드 압축 후 개수 세기.js
function solution(arr) {
  let answer = [0, 0];

  function compress(x, y, N) {
    let num = arr[x][y];
    for (let i = x; i < x + N; i++) {
      for (let j = y; j < y + N; j++) {
        if (arr[i][j] != num) {
          N = Math.floor(N / 2);
          compress(x, y, N);
          compress(x + N, y, N);
          compress(x, y + N, N);
          compress(x + N, y + N, N);
          return;
        }
      }
    }
    answer[num] += 1;
  }

  compress(0, 0, arr.length);
  return answer;
}

File: 프로그래머스/python/Level 2/쿼드 압축 후 개수 세기.py
def solution(arr):
    answer = [0, 0]
    
    def compress(x, y, N):
        num = arr[x][y]
        for i in range(x, x + N):
            for j in range(y, y + N):
                if arr[i][j] != num:
                    N = N // 2
                    compress(x, y, N)
                    compress(x+N, y, N)
                    compress(x, y+N, N)
                    compress(x+N, y+N, N)
                    return               
        answer[num] += 1  
        
    compress(0, 0, len(arr))                
    return answer